espresso/streamer.go

```diff
@@ -22,28 +18,27 @@ type L1Client interface {
 	HeaderByNumber(ctx context.Context, number *big.Int) (*types.Header, error)
 }
 
-type EspressoBatchI interface {
-	ToIncompleteBlock(rollupCfg *rollup.Config) (*types.Block, error)
-}
-
-type BatchBuffer interface {
-	Empty()
-	SetHeader(header espressoTypes.HeaderImpl)
-	SetBatchPos(pos uint64)
-	SetBatcherAddress(address common.Address)
-	ParseAndInsert(data []byte)
-	ReferenceL1BlockNumber() uint64
-	RemoveFirst()
-	Get(pos int) EspressoBatchI
-	Len() int
+// espresso-network-go's HeaderInterface currently lacks a function to get this info,
+// although it is present in all header versions
+func getFinalizedL1(header *espressoTypes.HeaderImpl) espressoTypes.L1BlockInfo {
+	v0_1, ok := header.Header.(*espressoTypes.Header0_1)
+	if ok {
+		return *v0_1.L1Finalized
+	}
+	v0_2, ok := header.Header.(*espressoTypes.Header0_2)
+	if ok {
+		return *v0_2.L1Finalized
+	}
+	v0_3, ok := header.Header.(*espressoTypes.Header0_3)
+	if ok {
+		return *v0_3.L1Finalized
+	}
+	panic("Unsupported header version")
 }
 
-type EspressoStreamer struct {
+type EspressoStreamer[B Batch] struct {
```

Comment:
Generic over batch type, so we can define a batch - which depends on types from op-node - in op-node, but everything else gets it's own package. Keeps op-node changes smaller.
Theoretically we could build on it and split it out into it's own repository later for future integrations to use

Reply:
Follow-up: https://app.asana.com/1/1208976916964769/project/1209392461754458/task/1209980878478920?focus=true